### PR TITLE
Map 113 - Improvement Task for WFS Controller

### PR DESCRIPTION
**Pull request by wchen329**
In this pull request:
* The WFS Web Service becomes an official THREDDS Web Service and it works correctly through THREDDS Catalog Web Service Access. In order to enable it add this service declaration to the catalog.xml:

`<service name="wfs" serviceType="WFS" base="/thredds/wfs/" />`

* The WFS Web Service no longer enforces all upper case parameter names or values. Any mix of casing (such as rEqUEST) is now accepted.

* The WFS Web Service now enforces the SERVICE and VERSION parameters. SERVICE requires value WFS and VERSION requires 2, 2.0, or 2.0.0, 2.0.2... etc. Of course, the values are not case sensitive.

* Enforcement of valid values for REQUEST as well (GetCapabilities DescribeFeature and GetFeature).

### CODE REVIEW

File: tds/src/main/java/thredds/core/StandardService.java [hunk 1]
```diff
@@ -26,6 +26,7 @@
   opendap(ServiceType.OPENDAP, "/dodsC/"),
   wcs(ServiceType.WCS, "/wcs/"),
   wms(ServiceType.WMS, "/wms/"),
+  wfs(ServiceType.WFS, "/wfs/"),
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/core/StandardService.java:29]: Nice find! I'm glad we can register our new service!

File: tds/src/main/java/thredds/server/wfs/WFSController.java [hunk 2]
```diff
@@ -49,14 +55,137 @@ private void getCapabilities(PrintWriter out, HttpServletRequest hsreq) {
...
+			String request = null;
+			String version = null;
+			String service = null;
+			
+			/* Look for parameter names to assign values
+			 * in order to avoid casing issues with parameter names (such as a mismatch between reQUEST and request and REQUEST).
+			 */
+			
+			for(String paramName : paramNames) {
+				if(paramName.equalsIgnoreCase("REQUEST")) {
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSController.java]: Good call on avoiding case sensitivity here and throughout the logic evaluating request parameters.

File: tds/src/main/java/thredds/server/wfs/WFSController.java [hunk 3]
```diff
@@ -49,14 +55,137 @@ private void getCapabilities(PrintWriter out, HttpServletRequest hsreq) {
...
+				// Only go through version checks if NOT a Get Capabilities request, the VERSION parameter is required for all operations EXCEPT GetCapabilities section 7.6.25 of WFS 2.0 Interface Standard
+				if(!request.equalsIgnoreCase(WFSRequestType.GetCapabilities.toString())) {
+				
+					if(version != null ) {
+						// If the version is not failed report exception VersionNegotiationFailed, from OGC Web Services Common Standard section 7.4.1
+						
+						List<String> versionParts = new ArrayList<String>();
+						String currentPart = "";
+						
+						//Interestingly, string split has problems with "." so split it manually
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSController.java]: I was surprised by this so I made a little experiment. Java's `String.split` method accepts a String regular expression as a parameter. Since the "." regular expression matches any character, the results can be confusing. Escaping the "." character to "\\\\." matches the literal period, so the results are much more straightforward.

```

class Main {
  public static void main(String[] args) {
    String[] versions = new String[] {
      "2.0.0",
      "2.0.2",
      "2.0",
      "2",
      //some edge cases
      ".",
      "..",
      ""
    };
    System.out.println("Listing all versions with their components split onto individual lines");
    for(String version : versions) {
      System.out.println("Splitting " + version);
      for(String versionComponent : version.split("\\.")) {
        System.out.println(versionComponent);
      }
    }
  System.out.println("Done listing all version components");
  }
}
```

```
Listing all versions with their components split onto individual lines
Splitting 2.0.0
2
0
0
Splitting 2.0.2
2
0
2
Splitting 2.0
2
0
Splitting 2
2
Splitting .
Splitting ..
Splitting 

Done listing all version components
```

You can play with this code snippet [here](https://repl.it/repls/BogusSoreDataset).

Does escaping the period eliminate the problems you observed before? Does this information change the way you would like to structure the version parsing and validation?

File: tds/src/main/java/thredds/server/wfs/WFSController.java [hunk 4]
```diff
@@ -49,14 +55,137 @@ private void getCapabilities(PrintWriter out, HttpServletRequest hsreq) {
...
+						for(int ind = 0; ind < version.length(); ind++) {
+							
+							// Check if number. If it is add it to the version part list
+							if(version.charAt(ind) >= '0' && version.charAt(ind) <= '9') {
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSController.java]: This is clever! If it doesn't work as well in a method that leverages String.split, you could consider using something like [Integer.parseUnsignedInt](https://docs.oracle.com/javase/8/docs/api/java/lang/Integer.html#parseUnsignedInt-java.lang.String-).

File: tds/src/main/java/thredds/server/wfs/WFSExceptionWriter.java [hunk 5]
```diff
@@ -5,6 +5,12 @@
 
 import javax.servlet.http.HttpServletResponse;
 
+/**
+ * A simple XML writer for XML-based WFS exceptions.
+ * 
+ * @author wchen@usgs.gov
+ *
+ */
 public class WFSExceptionWriter {
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSExceptionWriter.java:14]: Good call to make locators optional!

File: tds/src/main/java/thredds/server/wfs/WFSController.java [hunk 6]
```diff
@@ -49,14 +55,137 @@ private void getCapabilities(PrintWriter out, HttpServletRequest hsreq) {
 	@RequestMapping("**")
 	public void httpHandler(HttpServletRequest hsreq, HttpServletResponse hsres) {
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSController.java:146]: I like this comprehensive parameter parsing and validation. I like that several edge cases are addressed.

In general, the longer the method, the more difficult it is to test. In general, the longer the method, the more likely that the method is doing reasoning at multiple levels of abstraction, which can make it harder for a subsequent programmer to reason about the method. Since this method is getting longer, it is worth asking: How could this method be broken up into smaller, more focused methods?

At a hand-waving level, I've seen web service layer classes successfully keep their methods smaller and more focused by separating the following concerns: request parameter parsing, request parameter validation, request dispatching/execution.

You might draw some more concrete inspiration from the analogous [method (doGet) in WMSServlet](https://github.com/Reading-eScience-Centre/edal-java/blob/0ec0dd7ed3480c56913b91d4a026eaa420636507/wms/src/main/java/uk/ac/rdg/resc/edal/wms/WmsServlet.java#L288-L343) in edal-java, though you needn't strictly follow the approach that they use.

**Review by cschroedl-gov — CHANGES_REQUESTED** [review]
This is good stuff! I'm excited to see how the httpHandler could be split up a bit more. I'm curious if the code could be simplified using String.split with the escaped period literal.

File: tds/src/main/java/thredds/core/StandardService.java (hunk 1 above)
Comment (wchen329) [reply tds/src/main/java/thredds/core/StandardService.java:29]: Yeah! Becoming official

File: tds/src/main/java/thredds/server/wfs/WFSController.java (hunk 2 above)
Comment (wchen329) [reply tds/src/main/java/thredds/server/wfs/WFSController.java]: Yeah it was a little tricky since it required the deprecated Enumeration, but glad it could be fixed!

File: tds/src/main/java/thredds/server/wfs/WFSController.java (hunk 3 above)
Comment (wchen329) [reply tds/src/main/java/thredds/server/wfs/WFSController.java]: Interesting! I did try escaping the period previously with no effect. But perhaps that's because I did not escape escape it (i.e. `\.` instead of `\\.`). We could go with split I suppose although the current implementation is working as is.

File: tds/src/main/java/thredds/server/wfs/WFSController.java (hunk 4 above)
Comment (wchen329) [reply tds/src/main/java/thredds/server/wfs/WFSController.java]: Ha thanks! Yeah I don't forsee any problems with a split version but will keep that in mind.

File: tds/src/main/java/thredds/server/wfs/WFSExceptionWriter.java (hunk 5 above)
Comment (wchen329) [reply tds/src/main/java/thredds/server/wfs/WFSExceptionWriter.java:14]: Thanks! Yeah them standards.

File: tds/src/main/java/thredds/server/wfs/WFSController.java (hunk 6 above)
Comment (wchen329) [reply tds/src/main/java/thredds/server/wfs/WFSController.java:146]: Yeah, We could have the checks as a separate function.

File: tds/src/main/java/thredds/server/wfs/WFSController.java (hunk 3 above)
Comment (wchen329) [reply tds/src/main/java/thredds/server/wfs/WFSController.java]: This will be updated in a commit in the near future.

**Comment by wchen329** [comment]
Thanks for the feedback Carl! Alright so the most recent commit:
- WFS parameter checking was moved to separate method (is more granularity a goal)?
- Split is now used, different "methods" for checking correct formatting.
- Switch statement now used for checking request types.

File: tds/src/main/java/thredds/server/wfs/WFSController.java [hunk 7]
```diff
@@ -39,6 +45,96 @@ private void getCapabilities(PrintWriter out, HttpServletRequest hsreq) {
...
+		}
+					
+		else {
+			return new WFSExceptionWriter("WFS server error. SERVICE parameter is required.", "request", "MissingParameterValue");
+		}
+		
+		// The REQUEST Parameter is required. If not specified, is an error (throw exception through XML).
+		if(request != null) {
+			
+			// Only go through version checks if NOT a Get Capabilities request, the VERSION parameter is required for all operations EXCEPT GetCapabilities section 7.6.25 of WFS 2.0 Interface Standard
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSController.java:74]: Cool! I didn't know that `version` was optional with `GetCapabilities` requests. Good spec-checking!

File: tds/src/main/java/thredds/server/wfs/WFSRequestType.java [hunk 8]
```diff
@@ -28,4 +28,18 @@ public int toID()
...
 		return this.number;
 	}
+	
+	/**
+	 * Converts a string to a WFSRequestType if possible.
+	 * 
+	 * @param request value
+	 * @return the string as a WFSRequestType if possible.
+	 */
+	public static WFSRequestType getWFSRequestType(String request) {
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSRequestType.java:38]: Good idea to put this functionality in the enum class!

File: tds/src/main/java/thredds/server/wfs/WFSController.java [hunk 9]
```diff
@@ -39,6 +45,96 @@ private void getCapabilities(PrintWriter out, HttpServletRequest hsreq) {
...
+	/**
+	 * Checks request parameters for errors.
+	 * Will send back an XML Exception if any errors are encountered.
+	 * 
+	 * @param request parameter value
+	 * @param version parameter value
+	 * @param service parameter value
+	 * @return an ExceptionWriter if any errors occurred or null if none occurred
+	 */
+	private WFSExceptionWriter checkParametersForError(String request, String version, String service) {
```
Comment (cschroedl-gov) [inline tds/src/main/java/thredds/server/wfs/WFSController.java:57]: Good parameter validation throughout this method! When the validation finds that users have submitted invalid values, it might be helpful to include the invalid values in the WFS Exception responses. However this pull request is looking really good, so that isn't a blocker for merging.

**Review by cschroedl-gov — APPROVED** [review]
Looking good! No blockers here! Merge when ready.
